target_chains/ethereum/contracts/forge-test/PythGovernance.t.sol

```diff
@@ -0,0 +1,475 @@
+        );
+
+        bytes memory vaa = encodeAndSignMessage(
+            data,
+            TEST_GOVERNANCE_CHAIN_ID,
+            TEST_GOVERNANCE_EMITTER,
+            1
+        );
+
+        vm.expectRevert(PythErrors.InvalidGovernanceMessage.selector);
+        PythGovernance(address(pyth)).executeGovernanceInstruction(vaa);
+    }
+
+    function testInvalidGovernanceTarget() public {
+        // Test with wrong chain target
+        bytes memory data = abi.encodePacked(
+            MAGIC,
+            uint8(GovernanceModule.Target),
+            uint8(GovernanceAction.SetValidPeriod),
+            uint16(3), // Different chain ID for testing invalid target
+            uint64(0)
+        );
+
+        bytes memory vaa = encodeAndSignMessage(
+            data,
+            TEST_GOVERNANCE_CHAIN_ID,
+            TEST_GOVERNANCE_EMITTER,
+            1
+        );
+
+        vm.expectRevert(PythErrors.InvalidGovernanceTarget.selector);
+        PythGovernance(address(pyth)).executeGovernanceInstruction(vaa);
+    }
+
+    function testInvalidGovernanceDataSource() public {
+        // Test with wrong emitter
+        bytes memory data = abi.encodePacked(
+            MAGIC,
+            uint8(GovernanceModule.Target),
+            uint8(GovernanceAction.SetValidPeriod),
+            bytes32("ethereum"),
+            uint64(0)
+        );
+
+        bytes memory vaa = encodeAndSignMessage(
+            data,
+            TEST_GOVERNANCE_CHAIN_ID,
+            bytes32(uint256(0x1111)), // Wrong emitter
+            1
+        );
+
+        vm.expectRevert(PythErrors.InvalidGovernanceDataSource.selector);
+        PythGovernance(address(pyth)).executeGovernanceInstruction(vaa);
+    }
+
+    function testSetDataSources() public {
+        // Create governance VAA to set new data sources
+        bytes memory data = abi.encodePacked(
+            MAGIC,
+            uint8(GovernanceModule.Target),
+            uint8(GovernanceAction.SetDataSources),
+            TARGET_CHAIN_ID, // Target chain ID
+            uint8(1), // Number of data sources
+            uint16(1), // Chain ID
+            bytes32(
+                0x0000000000000000000000000000000000000000000000000000000000001111
+            ) // Emitter
+        );
+
+        bytes memory vaa = encodeAndSignMessage(
+            data,
+            TEST_GOVERNANCE_CHAIN_ID,
+            TEST_GOVERNANCE_EMITTER,
+            1
+        );
+
+        PythInternalStructs.DataSource[] memory oldDataSources = PythGetters(
+            address(pyth)
+        ).validDataSources();
+
+        PythInternalStructs.DataSource[]
+            memory newDataSources = new PythInternalStructs.DataSource[](1);
+        newDataSources[0] = PythInternalStructs.DataSource(
+            1,
+            0x0000000000000000000000000000000000000000000000000000000000001111
+        );
+
+        vm.expectEmit(true, true, true, true);
+        emit DataSourcesSet(oldDataSources, newDataSources);
```

Comment:
why is this here? i see it in many other places. we expect the call below to emit them not manually emitting them.

Reply:
this is not manually emitting them, docs [here](https://book.getfoundry.sh/cheatcodes/expect-emit)

but basically whats happening is:
1. vm.expectEmit(true, true, true, true) tells Forge to expect an event where:
- Each true corresponds to checking topic1, topic2, topic3, and data respectively
- All true means "check everything in the event"
2. The subsequent emit defines the EXPECTED event parameters
3. When executeGovernanceInstruction(vaa) runs, Forge compares its emitted event against this expected template

This pattern is used because we can't directly capture and inspect events after they're emitted. Instead, we declare what you expect before the action, then Forge verifies the actual emission matches.

Reply:
ah my bad here. thanks.